java/ql/src/semmle/code/java/frameworks/android/SQLite.qll

```diff
@@ -226,3 +227,73 @@ private class ContentProviderUpdateMethod extends SQLiteRunner {
 
   override int sqlIndex() { result = 2 }
 }
+
+private class QueryBuilderBuildMethod extends TaintPreservingMethod {
+  QueryBuilderBuildMethod() {
```

Comment:
Have you considered using a field like `int argument` and write this like:
```
private class QueryBuilderBuildMethod extends TaintPreservingMethod {
  int argument;

  QueryBuilderBuildMethod() {
    this.getDeclaringType().getASourceSupertype*() instanceof Class and
    // buildQuery(String[] projectionIn, String selection, String groupBy, String having, String sortOrder, String limit)
    // buildQuery(String[] projectionIn, String selection, String[] selectionArgs, String groupBy, String having, String sortOrder, String limit)
    // buildUnionQuery(String[] subQueries, String sortOrder, String limit)
    // buildUnionSubQuery(String typeDiscriminatorColumn, String[] unionColumns, Set<String> columnsPresentInTable, int computedColumnsOffset, String typeDiscriminatorValue, String selection, String[] selectionArgs, String groupBy, String having)
    // buildUnionSubQuery(String typeDiscriminatorColumn, String[] unionColumns, Set<String> columnsPresentInTable, int computedColumnsOffset, String typeDiscriminatorValue, String selection, String groupBy, String having)
    // static buildQueryString(boolean distinct, String tables, String[] columns, String where, String groupBy, String having, String orderBy, String limit)
    this.hasName(["buildQuery", "buildUnionQuery", "buildUnionSubQuery", "buildQueryString"]) and
    argument = -1
    or
    hasName(["buildQuery", "buildUnionQuery"]) and
    argument = [0 .. getNumberOfParameters()]
    or
    hasName("buildQueryString") and
    argument = [1 .. getNumberOfParameters()]
    or
    hasName("buildUnionSubQuery") and
    argument = [0 .. getNumberOfParameters()] and
    argument != 3
  }

  override predicate returnsTaint(int arg) { argument = arg }
}
```

This way you don't need to split the implementation over two predicates.

Reply:
Good suggestion